scrypto/src/resource/resource_builder.rs

```diff
@@ -1,549 +1,489 @@
+    ) -> InProgressResourceBuilder<NonFungibleResourceType<BytesNonFungibleLocalId>, NoAuth> {
+        InProgressResourceBuilder::default()
     }
 
     /// Starts a new builder to create a non-fungible resource with a `NonFungibleIdType::UUID`
-    pub fn new_uuid_non_fungible() -> NonFungibleResourceBuilder<UUIDNonFungibleLocalId> {
-        NonFungibleResourceBuilder::new()
+    pub fn new_uuid_non_fungible(
+    ) -> InProgressResourceBuilder<NonFungibleResourceType<UUIDNonFungibleLocalId>, NoAuth> {
+        InProgressResourceBuilder::default()
     }
 }
 
-/// A resource builder which builds fungible resources that may or may not have an owner badge.
+/// Utility for setting up a new resource, which has building in progress.
 ///
-/// # Note
+/// * You start the building process with one of the methods starting with `ResourceBuilder::new_`.
+/// * The allowed methods change depending on which methods have already been called.
+///   For example, you can either use `owner_non_fungible_badge` or set access rules individually, but not both.
+/// * You can complete the building process using either `create_with_no_initial_supply()` or `mint_initial_supply(..)`.
 ///
-/// Once one of the methods that set behavior is called, a new [`FungibleResourceWithAuthBuilder`]
-/// is created which commits the developer to building a resource that does not have an owner badge.
-/// If none of these methods are called, then the developer has the choice to either building a
-/// resource with an owner badge or without one.
-pub struct FungibleResourceBuilder {
-    divisibility: u8,
+/// ### Example
+/// ```no_run
+/// use scrypto::prelude::*;
+///
+/// let bucket = ResourceBuilder::new_fungible()
+///     .metadata("name", "TestToken")
+///     .mint_initial_supply(5);
+/// ```
+#[must_use]
+pub struct InProgressResourceBuilder<T: ResourceType, A: ConfiguredAuth> {
+    resource_type: T,
     metadata: BTreeMap<String, String>,
+    auth: A,
 }
 
-impl FungibleResourceBuilder {
-    pub fn new() -> Self {
+impl<T: ResourceType> Default for InProgressResourceBuilder<T, NoAuth> {
+    fn default() -> Self {
         Self {
-            divisibility: DIVISIBILITY_MAXIMUM,
+            resource_type: T::default(),
             metadata: BTreeMap::new(),
+            auth: NoAuth,
         }
     }
+}
 
-    /// Set the divisibility.
-    ///
-    /// `0` means the resource is not divisible; `18` is the max divisibility.
-    pub fn divisibility(mut self, divisibility: u8) -> Self {
-        assert!(divisibility <= 18);
-        self.divisibility = divisibility;
-        self
-    }
-
-    /// Adds a resource metadata.
-    ///
-    /// If a previous attribute with the same name has been set, it will be overwritten.
-    pub fn metadata<K: AsRef<str>, V: AsRef<str>>(mut self, name: K, value: V) -> Self {
-        self.metadata
-            .insert(name.as_ref().to_owned(), value.as_ref().to_owned());
-        self
-    }
+pub trait ConfiguredAuth {
```

Comment:
Why not just `Into<BTreeMap<ResourceMethodAuthKey, (AccessRule, AccessRule)>>`?

Reply:
I actually tried it briefly and it was just uglier to implement / more lines of code :D.

Both are fine obviously.